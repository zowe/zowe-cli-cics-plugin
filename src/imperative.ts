/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

// Imperative version of Zowe CLI
import { IImperativeConfig } from "@brightside/imperative";
import { PluginConstants } from "./api/constants/PluginConstants";
import { CicsSession } from "./cli/CicsSession";

const config: IImperativeConfig = {
    commandModuleGlobs: ["**/cli/*/*.definition!(.d).*s"],
    rootCommandDescription: PluginConstants.PLUGIN_DESCRIPTION,
    productDisplayName: PluginConstants.PLUGIN_NAME,
    name: PluginConstants.PLUGIN_GROUP_NAME,
    pluginHealthCheck: "./lib/healthCheck.handler",
    profiles: [
        {
            type: "cics",
            schema: {
                type: "object",
                title: "CICS Profile",
                description: "A cics profile is required to issue commands in the cics command group that interact with " +
                    "CICS regions. The cics profile contains your host, port, user name, and password " +
                    "for the IBM CICS management client interface (CMCI) server of your choice.",
                properties: {
                    host: {
                        type: "string",
                        optionDefinition: {
                            name: "host",
                            aliases: ["H"],
                            description: "The CMCI server host name",
                            type: "string",
                            required: true,
                        },
                    },
                    port: {
                        type: "number",
                        optionDefinition: {
                            name: "port",
                            aliases: ["P"],
                            description: "The CMCI server port",
                            type: "number",
                            defaultValue: 1490,
                        },
                    },
                    user: {
                        type: "string",
                        secure: true,
                        optionDefinition: {
                            name: "user",
                            aliases: ["u"],
                            description: "Your username to connect to CICS",
                            type: "string",
                            implies: ["password"],
                            required: true,
                        },
                    },
                    password: {
                        type: "string",
                        secure: true,
                        optionDefinition: {
                            name: "password",
                            aliases: ["p"],
                            description: "Your password to connect to CICS",
                            type: "string",
                            implies: ["user"],
                            required: true,
                        },
                    },
                    regionName: {
                        type: "string",
                        optionDefinition: {
                            name: "region-name",
                            description: "The name of the CICS region name to interact with",
                            type: "string"
                        },
                    },
                    cicsPlex: {
                        type: "string",
                        optionDefinition: {
                            name: "cics-plex",
                            description: "The name of the CICSPlex to interact with",
                            type: "string"
                        },
                    },
                    rejectUnauthorized: {
                        type: "boolean",
                        optionDefinition: CicsSession.CICS_OPTION_REJECT_UNAUTHORIZED
                    },
                    protocol: {
                        type: "string",
                        optionDefinition: CicsSession.CICS_OPTION_PROTOCOL
                    }
                },
                required: ["host"],
            },
            createProfileExamples: [
                {
                    options: "cics123 --host zos123 --port 1490 --user ibmuser --password myp4ss",
                    description: "Create a cics profile named 'cics123' to connect to CICS at host zos123 and port 1490"
                }
            ]
        }
    ]
};

export = config;
