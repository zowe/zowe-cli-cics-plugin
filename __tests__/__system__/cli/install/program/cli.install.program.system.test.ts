/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import { TestEnvironment } from "../../../../__src__/environment/TestEnvironment";
import { ITestEnvironment } from "../../../../__src__/environment/doc/response/ITestEnvironment";
import { generateRandomAlphaNumericString, runCliScript } from "../../../../__src__/TestUtils";
import { Session } from "@brightside/imperative";
import { CicsCmciConstants, CicsCmciRestClient } from "../../../../../src";

let TEST_ENVIRONMENT: ITestEnvironment;
let regionName: string;
let csdGroup: string;
let host: string;
let port: number;
let user: string;
let password: string;
let protocol: string;
let rejectUnauthorized: boolean;

describe("CICS install program command", () => {

    beforeAll(async () => {
        TEST_ENVIRONMENT = await TestEnvironment.setUp({
            testName: "install_program",
            installPlugin: true,
            tempProfileTypes: ["cics"]
        });
        csdGroup = TEST_ENVIRONMENT.systemTestProperties.cmci.csdGroup;
        regionName = TEST_ENVIRONMENT.systemTestProperties.cmci.regionName;
        host = TEST_ENVIRONMENT.systemTestProperties.cmci.host;
        port = TEST_ENVIRONMENT.systemTestProperties.cmci.port;
        user = TEST_ENVIRONMENT.systemTestProperties.cmci.user;
        password = TEST_ENVIRONMENT.systemTestProperties.cmci.password;
        protocol = TEST_ENVIRONMENT.systemTestProperties.cmci.protocol;
        rejectUnauthorized = TEST_ENVIRONMENT.systemTestProperties.cmci.rejectUnauthorized;
    });

    afterAll(async () => {
        await TestEnvironment.cleanUp(TEST_ENVIRONMENT);
    });

    const deleteProgram = async (programName: string) => {
        const cmciProperties = TEST_ENVIRONMENT.systemTestProperties.cmci;
        const session = new Session({
            type: "basic",
            hostname: cmciProperties.host,
            port: cmciProperties.port,
            user: cmciProperties.user,
            password: cmciProperties.password,
            strictSSL: false,
            protocol: protocol as any || "http",
        });

        return CicsCmciRestClient.deleteExpectParsedXml(session,
            `/${CicsCmciConstants.CICS_SYSTEM_MANAGEMENT}/CICSDefinitionProgram/${cmciProperties.regionName}` +
            `?CRITERIA=(NAME=${programName})&PARAMETER=CSDGROUP(${cmciProperties.csdGroup})`);
    };

    const discardProgram = async (programName: string) => {
        const cmciProperties = TEST_ENVIRONMENT.systemTestProperties.cmci;
        const deleteSession = new Session({
            type: "basic",
            hostname: cmciProperties.host,
            port: cmciProperties.port,
            user: cmciProperties.user,
            password: cmciProperties.password,
            strictSSL: false,
            protocol: protocol as any || "http",
        });

        return CicsCmciRestClient.deleteExpectParsedXml(deleteSession,
            `/${CicsCmciConstants.CICS_SYSTEM_MANAGEMENT}/CICSProgram/${cmciProperties.regionName}` +
            `?CRITERIA=(PROGRAM=${programName})`);
    };

    it("should be able to display the help", () => {
        const output = runCliScript(__dirname + "/__scripts__/install_program_help.sh", TEST_ENVIRONMENT, []);
        expect(output.stderr.toString()).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toMatchSnapshot();
    });

    it("should be able to successfully install a program with basic options", async () => {

        // first define the program
        const programNameSuffixLength = 4;
        const programName = "AAA" + generateRandomAlphaNumericString(programNameSuffixLength);
        let output = runCliScript(__dirname + "/../../define/program/__scripts__/define_program.sh", TEST_ENVIRONMENT,
            [programName, csdGroup, regionName]);
        let stderr = output.stderr.toString();
        expect(stderr).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toContain("success");

        output = runCliScript(__dirname + "/__scripts__/install_program.sh", TEST_ENVIRONMENT, [programName, csdGroup, regionName]);
        stderr = output.stderr.toString();
        expect(stderr).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toContain("success");
        await discardProgram(programName);
        await deleteProgram(programName);
    });

    it("should get a syntax error if program name is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/install_program.sh", TEST_ENVIRONMENT, ["", "FAKEGRP", "FAKERGN"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("program");
        expect(stderr).toContain("name");
        expect(output.status).toEqual(1);
    });

    it("should get a syntax error if CSD group is omitted", () => {
        const output = runCliScript(__dirname + "/__scripts__/install_program.sh", TEST_ENVIRONMENT, ["FAKEPGM", "", "FAKERGN"]);
        const stderr = output.stderr.toString();
        expect(stderr).toContain("Syntax");
        expect(stderr).toContain("csdGroup");
        expect(output.status).toEqual(1);
    });

    it("should be able to successfully install a program with profile options", async () => {

        // first define the program
        const programNameSuffixLength = 4;
        const programName = "AAA" + generateRandomAlphaNumericString(programNameSuffixLength);
        let output = runCliScript(__dirname + "/../../define/program/__scripts__/define_program_fully_qualified.sh", TEST_ENVIRONMENT,
            [programName,
                csdGroup,
                regionName,
                host,
                port,
                user,
                password,
                protocol,
                rejectUnauthorized]);
        let stderr = output.stderr.toString();
        expect(stderr).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toContain("success");

        output = runCliScript(__dirname + "/__scripts__/install_program_fully_qualified.sh", TEST_ENVIRONMENT,
            [programName,
                csdGroup,
                regionName,
                host,
                port,
                user,
                password,
                protocol,
                rejectUnauthorized]);
        stderr = output.stderr.toString();
        expect(stderr).toEqual("");
        expect(output.status).toEqual(0);
        expect(output.stdout.toString()).toContain("success");
        await discardProgram(programName);
        await deleteProgram(programName);
    });

});
