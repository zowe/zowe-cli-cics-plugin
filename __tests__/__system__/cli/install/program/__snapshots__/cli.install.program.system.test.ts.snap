// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CICS install program command should be able to display the help 1`] = `
"
 COMMAND NAME
 ------------

   program | prog

 DESCRIPTION
 -----------

   Install a program to CICS.

 USAGE
 -----

   zowe cics install program <programName> <csdGroup> [options]

 POSITIONAL ARGUMENTS
 --------------------

   programName		 (string)

      The name of the program to install. The maximum length of the program name is
      eight characters.

   csdGroup		 (string)

      The CICS system definition (CSD) Group for the program that you want to install.
      The maximum length of the group name is eight characters.

 OPTIONS
 -------

   --region-name  (string)

      The CICS region name to which to install the program

   --cics-plex  (string)

      The name of the CICSPlex to which to install the program

 CICS CONNECTION OPTIONS
 -----------------------

   --host  | -H (string)

      The CICS server host name.

   --port  | -P (number)

      The CICS server port.

      Default value: 443

   --user  | -u (string)

      Mainframe (CICS) user name, which can be the same as your TSO login.

   --password  | --pw (string)

      Mainframe (CICS) password, which can be the same as your TSO password.

   --reject-unauthorized  | --ru (boolean)

      Reject self-signed certificates.

      Default value: true

   --protocol  | -o (string)

      Specifies CMCI protocol (http or https).

      Default value: http
      Allowed values: http, https

 PROFILE OPTIONS
 ---------------

   --cics-profile  | --cics-p (string)

      The name of a (cics) profile to load for this command execution.

 GLOBAL OPTIONS
 --------------

   --response-format-json  | --rfj (boolean)

      Produce JSON formatted data from a command

   --help  | -h (boolean)

      Display help text

   --help-web  | --hw (boolean)

      Display HTML help in browser

 EXAMPLES
 --------

   - Install a program named PGM123 to the region named MYREGION
   in the CSD group MYGRP:

      $ zowe cics install program PGM123 MYGRP --region-name MYREGION

"
`;
